### PR TITLE
expression: rewrite builtin function: WEEK, WEEKOFYEAR

### PR DESCRIPTION
**Pull request by zz-jason**
to #4080

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -896,51 +897,71 @@ func (c *weekFunctionClass) getFunction(args []Expression, ctx context.Context)
...
-func (b *builtinWeekSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+// evalInt evals WEEK(date, mode).
+// see: https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_week
+func (b *builtinWeekWithModeSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	date, isNull, err := b.args[0].EvalTime(row, sc)
 	if err != nil {
-		return d, errors.Trace(err)
+		sc.AppendWarning(err)
```
Comment (shenli) [inline expression/builtin_time.go]: The old function will return immediately when meet error.

Comment (zz-jason) [inline expression/builtin_time.go]: it's a bug fix, to be compatible with mysql

Comment (shenli) [inline expression/builtin_time.go]: We should consider whether it is in a read or write statement. We could leave a TODO here.
mysql> SELECT WEEK('2008-xx',0);
+-------------------+
| WEEK('2008-xx',0) |
+-------------------+
|              NULL |
+-------------------+
1 row in set, 1 warning (0.00 sec)

mysql> insert into t values(WEEK('2008-xx',0));
ERROR 1292 (22007): Incorrect datetime value: '2008-xx'
mysql>

Comment (shenli) [inline expression/builtin_time.go]: And please add a test case for this.

**Comment by XuHuaiyu** [comment]
ci failed

File: expression/builtin_time.go (hunk 1 above)
Comment (zz-jason) [inline expression/builtin_time.go]: done

**Comment by zz-jason** [comment]
@XuHuaiyu @shenli PTAL

File: expression/integration_test.go [hunk 2]
```diff
@@ -936,6 +936,32 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
...
+	_, err := tk.Exec(`insert into t select week("aa", 1)`)
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Matches, "invalid time format")
+	tk.MustExec(`insert into t select now()`)
+	_, err = tk.Exec(`update t set a = week("aa", 1)`)
+	c.Assert(err.Error(), Matches, "invalid time format")
+	_, err = tk.Exec(`delete from t where a = week("aa", 1)`)
+	c.Assert(err.Error(), Matches, "invalid time format")
+
+	// for weekofyear
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: why so many test cases for week, but fewer for weekofyear?

File: expression/integration_test.go [hunk 3]
```diff
@@ -936,6 +936,42 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
...
+	result.Check(testkit.Rows("7 8 7 7 8"))
+	result = tk.MustQuery(`select week("aa", 1), week(null, 2), week(11, 2), week(12.99, 2);`)
+	result.Check(testkit.Rows("<nil> <nil> <nil> <nil>"))
+	result = tk.MustQuery(`select week("aa"), week(null), week(11), week(12.99);`)
+	result.Check(testkit.Rows("<nil> <nil> <nil> <nil>"))
+	tk.MustExec(`drop table if exists t`)
+	tk.MustExec(`create table t(a datetime)`)
+	_, err := tk.Exec(`insert into t select week("aa", 1)`)
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Matches, "invalid time format")
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: use 	c.Assert(terror.ErrorEqual(err, types.ErrInvalidTimeFormat), IsTrue)

File: expression/integration_test.go [hunk 4]
```diff
@@ -936,6 +936,42 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
...
+	tk.MustExec(`insert into t select now()`)
+	_, err = tk.Exec(`update t set a = week("aa", 1)`)
+	c.Assert(err.Error(), Matches, "invalid time format")
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: Will this error be ignored by constant fold?

Comment (zz-jason) [inline expression/integration_test.go]: error will be recorded to the log as a warning and stop folding the current expression.

**Comment by breeswish** [comment]
LGTM

**Pull request merged**